### PR TITLE
Added Attachments support for LRSs

### PR DESCRIPTION
**Pull request by FishHooks**
Rebased with current master and fixed tests to support version 1.0.2

Addresses #70 

This PR includes the following:

- Attachments can now be added to statements and sent to the LRS

- The attachments' content is sent with the statement but not serialized with the rest of the statement's JSON

- The LRS sends the put request in the correct format if the statement has an attachment (and the correct `Content-Type`)

- If the attachment flag is set to `true` in `queryStatements()`, the response is formatted correctly (previously `queryStatements()` did not handle multipart requests, and would therefore not return the attachments' content)

### CODE REVIEW

**Comment by brianjmiller** [comment]
It would be good to test the additional methods of LRS, and to have a test that sends an attachment, one that retrieves a statement via id, and one that query's statements.

File: src/Attachment.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
+/*
+    Copyright 2014 Rustici Software
```
Comment (brianjmiller) [inline src/Attachment.js]: New files should have updated year, 2016.

File: src/Attachment.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            ;
+
+            cfg = cfg || {};
+
+            if (cfg.hasOwnProperty("description") && typeof cfg.description !== "undefined" && cfg.description !== null) {
+                this.description = cfg.description;
+            }
+            if (cfg.hasOwnProperty("fileUrl") && typeof cfg.fileUrl !== "undefined" && cfg.fileUrl !== null) {
+                this.fileUrl = cfg.fileUrl;
+            }
```
Comment (brianjmiller) [inline src/Attachment.js]: Can't these two be `directProps`?

File: src/Attachment.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+            for (i = 0; i < directProps.length; i += 1) {
+                if (cfg.hasOwnProperty(directProps[i]) && cfg[directProps[i]] !== null) {
+                    this[directProps[i]] = cfg[directProps[i]];
+                }
+            }
+
+            if (cfg.hasOwnProperty("content") && typeof cfg.content !== "undefined" && cfg.content !== null) {
```
Comment (brianjmiller) [inline src/Attachment.js]: Not sure why we are checking both `hasOwnProperty` and the `typeof`, generally the latter is used to check to see if a property exists. Shouldn't need both.

File: src/Attachment.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        @type String
+        */
+        this.sha2 = null;
+
+        /**
+        @property description
+        @type Object
+        */
+        this.description = null;
+
```
Comment (brianjmiller) [inline src/Attachment.js]: Should include `content` in the prototype since it will sometimes be set, then can just check for `null` to see if it *has* been set.

File: src/Attachment.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+                result = {
+                    contentType: this.contentType,
+                    description: this.description,
+                    display: this.display,
+                    fileUrl: this.fileUrl,
+                    length: this.length,
+                    sha2: this.sha2,
+                    usageType: this.usageType
+                };
+                if (this.hasOwnProperty("content") && typeof this.content !== "undefined" && this.content !== null) {
```
Comment (brianjmiller) [inline src/Attachment.js]: `this` will *always* have `content` based on the object's prototype based on comment above, this can be reduced to just check for null.

File: src/Attachment.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            else {
+                result = {
+                    contentType: this.contentType,
+                    description: this.description,
+                    display: this.display,
+                    fileUrl: this.fileUrl,
+                    length: this.length,
+                    sha2: this.sha2,
+                    usageType: this.usageType
+                };
```
Comment (brianjmiller) [inline src/Attachment.js]: `description` and `fileUrl` are optional, need a `null` check to not include them.

File: src/LRS.js [hunk 7]
```diff
@@ -158,6 +158,39 @@ TinCan client library
         },
 
         /**
+        Creates and returns a boundary for separating parts in
+        requests where the statement has an attachment
+
+        @method getBoundary
+        @private
```
Comment (brianjmiller) [inline src/LRS.js]: JS doesn't have actual private/public, etc. so properties/methods that are "private"ish should start with `_`, please rename these.

File: src/LRS.js [hunk 8]
```diff
@@ -158,6 +158,39 @@ TinCan client library
...
+
+        /**
+        Returns the stringified version, with boundary and content-type,
+        of a statement to place in the request
+
+        @method createStatementSegment
+        @private
+        */
+        createStatementSegment: function (boundary, statement) {
+            return "--" + boundary + "\r\n" + "Content-Type:application/json" + "\r\n\r\n" + JSON.stringify(statement) + "\r\n";
```
Comment (brianjmiller) [inline src/LRS.js]: Per RFC a space is preferred after header names and before the value.

File: src/LRS.js [hunk 9]
```diff
@@ -310,7 +343,11 @@ TinCan client library
         saveStatement: function (stmt, cfg) {
             this.log("saveStatement");
             var requestCfg,
-                versionedStatement;
+                versionedStatement,
+                boundary,
+                flag = true,
+                attachmentContent = [],
+                a;
```
Comment (brianjmiller) [inline src/LRS.js]: Getting *really* nitpicky now :-), but `i` is more canonical in the lib (AFAICT all for loops use it), so let's switch for consistency.

File: src/LRS.js [hunk 10]
```diff
@@ -310,7 +343,11 @@ TinCan client library
         saveStatement: function (stmt, cfg) {
             this.log("saveStatement");
             var requestCfg,
-                versionedStatement;
+                versionedStatement,
+                boundary,
+                flag = true,
```
Comment (brianjmiller) [inline src/LRS.js]: `flag` for what? Better name please.

File: src/LRS.js [hunk 11]
```diff
@@ -341,13 +378,43 @@ TinCan client library
...
+            if (versionedStatement.hasOwnProperty("attachments") && versionedStatement.attachments !== null) {
+                boundary = this.getBoundary();
+                requestCfg = {
+                    url: "statements",
+                    headers: {
+                        "Content-Type": "application/json"
+                    }
+                };
+                for (a = 0; a < versionedStatement.attachments.length; a += 1) {
+                    if (typeof versionedStatement.attachments[a].content !== "undefined" && versionedStatement.attachments[a].content !== null) {
```
Comment (brianjmiller) [inline src/LRS.js]: Should be just `null` check after above.

File: src/LRS.js [hunk 12]
```diff
@@ -341,13 +378,43 @@ TinCan client library
...
+                        attachmentContent.push(versionedStatement.attachments[a].content);
+                        delete versionedStatement.attachments[a].content;
```
Comment (brianjmiller) [inline src/LRS.js]: This begs the question of whether the versioned statement should *ever* have the `content` set, since it can't be used in that way. It probably shouldn't, and instead we use the index of the versioned attachments to index into the original statement to check for content. See PHP for an example.

File: src/LRS.js [hunk 13]
```diff
@@ -341,13 +378,43 @@ TinCan client library
...
+                    requestCfg.data = this.createStatementSegment(boundary, versionedStatement);
+                    for (a = 0; a < attachmentContent.length; a += 1) {
+                        if (attachmentContent[a] !== null) {
+                            requestCfg.data += this.createAttachmentSegment(boundary, attachmentContent[a], versionedStatement.attachments[a].sha2, versionedStatement.attachments[a].contentType);
+                        }
+                    }
+                    requestCfg.data += "--" + boundary + "--";
+                }
+            }
+            else if (flag) {
```
Comment (brianjmiller) [inline src/LRS.js]: When would `flag` be `false`?

File: src/LRS.js [hunk 14]
```diff
@@ -542,6 +609,46 @@ TinCan client library
...
+
+            for (i = 0; i < lines.length; i += 1) {
+                if (lines[i].startsWith("X-Experience-API-Hash")) {
+                    hashToMatch = lines[i].split(":")[1];
+                }
+                else if (lines[i] === "") {
+                    i += 1;
+                    content = lines[i];
+                    for (x = 0; x < stmts.statements.length; x += 1) {
+                        if (stmts.statements[x].hasOwnProperty("attachments") && stmts.statements[x].attachments !== null) {
```
Comment (brianjmiller) [inline src/LRS.js]: Should be `null` only check.

File: src/LRS.js [hunk 15]
```diff
@@ -542,6 +609,46 @@ TinCan client library
...
+                x,
+                a;
+
+            for (i = 0; i < lines.length; i += 1) {
+                if (lines[i].startsWith("X-Experience-API-Hash")) {
+                    hashToMatch = lines[i].split(":")[1];
+                }
+                else if (lines[i] === "") {
+                    i += 1;
+                    content = lines[i];
```
Comment (brianjmiller) [inline src/LRS.js]: What happens if content is on more than one line?

File: src/LRS.js [hunk 16]
```diff
@@ -542,6 +609,46 @@ TinCan client library
...
+        _assignAttachmentContent: function (stmts, lines) {
+            var hashToMatch,
+                content,
+                i,
+                x,
+                a;
+
+            for (i = 0; i < lines.length; i += 1) {
+                if (lines[i].startsWith("X-Experience-API-Hash")) {
+                    hashToMatch = lines[i].split(":")[1];
```
Comment (brianjmiller) [inline src/LRS.js]: Not sure this is sufficient to parse headers. Can we avoid parsing the headers ourselves? At the very least this probably breaks on whitespace.

File: src/LRS.js [hunk 17]
```diff
@@ -604,10 +711,29 @@ TinCan client library
...
+                        i;
 
                     if (err === null) {
-                        result = TinCan.StatementsResult.fromJSON(xhr.responseText);
+                        if (!cfg.params.attachments) {
+                            result = TinCan.StatementsResult.fromJSON(xhr.responseText);
+                        }
+                        else {
+                            for (i = 0; i < responseLines.length; i += 1) {
+                                if (responseLines[i].startsWith("Content-Type:application/json")) {
```
Comment (brianjmiller) [inline src/LRS.js]: In sufficient for header detection.

File: src/LRS.js [hunk 18]
```diff
@@ -604,10 +711,29 @@ TinCan client library
...
+                                    i += 2;
```
Comment (brianjmiller) [inline src/LRS.js]: ?? No guarantee content type is the last header.

File: src/Statement.js [hunk 19]
```diff
@@ -179,7 +186,8 @@ TinCan client library
                     "version",
                     "inProgress",
                     "voided"
-                ];
+                ],
+                a;
```
Comment (brianjmiller) [inline src/Statement.js]: prefer `i` for single indexing.

File: src/Statement.js [hunk 20]
```diff
@@ -278,6 +286,16 @@ TinCan client library
                     this.context = new TinCan.Context (cfg.context);
                 }
             }
+            if (cfg.hasOwnProperty("attachments") && cfg.attachments !== null && typeof cfg.attachments !== "undefined") {
```
Comment (brianjmiller) [inline src/Statement.js]: Don't need both checks.

File: src/Statement.js [hunk 21]
```diff
@@ -278,6 +286,16 @@ TinCan client library
...
+                this.attachments = [];
+                for (a = 0; a < cfg.attachments.length; a += 1) {
+                    if (!(cfg.attachments[a] instanceof TinCan.Attachment)) {
+                        this.attachments.push(new TinCan.Attachment (cfg.attachments[a]));
+                    } else {
```
Comment (brianjmiller) [inline src/Statement.js]: Eek, cuddled else!

File: src/LRS.js [hunk 22]
```diff
@@ -618,7 +744,7 @@ TinCan client library
             requestResult = this.sendRequest(requestCfg);
             requestResult.config = requestCfg;
 
-            if (! callbackWrapper) {
+            if (!callbackWrapper) {
```
Comment (brianjmiller) [inline src/LRS.js]: Should be a space here.

File: src/LRS.js [hunk 23]
```diff
@@ -604,10 +711,29 @@ TinCan client library
...
                 callbackWrapper = function (err, xhr) {
-                    var result = xhr;
+                    var result = xhr,
+                        responseLines = xhr.responseText.split("\r\n"),
+                        statements,
+                        i;
 
                     if (err === null) {
-                        result = TinCan.StatementsResult.fromJSON(xhr.responseText);
+                        if (!cfg.params.attachments) {
```
Comment (brianjmiller) [inline src/LRS.js]: Space after operator.

File: src/LRS.js [hunk 24]
```diff
@@ -341,13 +378,43 @@ TinCan client library
...
+                        attachmentContent.push(versionedStatement.attachments[a].content);
+                        delete versionedStatement.attachments[a].content;
+                    }
                 }
-            };
+                if (attachmentContent.length !== 0) {
+                    flag = false;
+                    requestCfg.headers["Content-Type"] = "multipart/mixed; boundary=" + boundary;
+                }
+                if (!flag) {
```
Comment (brianjmiller) [inline src/LRS.js]: Space after operator.

File: src/Statement.js [hunk 25]
```diff
@@ -278,6 +286,16 @@ TinCan client library
                     this.context = new TinCan.Context (cfg.context);
                 }
             }
+            if (cfg.hasOwnProperty("attachments") && cfg.attachments !== null && typeof cfg.attachments !== "undefined") {
+                this.attachments = [];
+                for (a = 0; a < cfg.attachments.length; a += 1) {
+                    if (!(cfg.attachments[a] instanceof TinCan.Attachment)) {
```
Comment (brianjmiller) [inline src/Statement.js]: Space after operator.

File: src/Statement.js [hunk 26]
```diff
@@ -321,7 +339,9 @@ TinCan client library
                     "context",
                     "authority"
                 ],
-                i;
+                a,
+                i,
+                temp;
```
Comment (brianjmiller) [inline src/Statement.js]: `temp` should be unnecessary, use method chaining.

Comment (brianjmiller) [inline src/Statement.js]: Alternatively, always set it (as a better name) and drop the `.push` redundancy.

File: test/js/unit/Attachment.js [hunk 27]
```diff
@@ -0,0 +1,167 @@
+/*
+    Copyright 2014 Rustici Software
```
Comment (brianjmiller) [inline test/js/unit/Attachment.js]: +1

File: test/js/unit/Attachment.js [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+            ;
+            ok(obj instanceof TinCan.Attachment, "object is TinCan.Attachment");
+
+            for(i = 0; i < nullProps.length; i += 1) {
+                ok(obj.hasOwnProperty(nullProps[i]), "object has property: " + nullProps[i]);
+                strictEqual(obj[nullProps[i]], null, "object property initial value: " + nullProps[i]);
+            }
+
+            strictEqual(obj.LOG_SRC, "Attachment", "object property LOG_SRC initial value");
+            //strictEqual(obj.objectType, "Attachment", "object property objectType initial value");
```
Comment (brianjmiller) [inline test/js/unit/Attachment.js]: Remove.

File: test/js/unit/Attachment.js [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+                    name: "Attachment with content",
+                    instanceConfig: {
+                        content: "test text content"
+                    },
+                    checkProps: {
+                        content: "test text content"
+                    }
+                }
+            ],
+                i,
```
Comment (brianjmiller) [inline test/js/unit/Attachment.js]: Something is odd with this indenting.

File: test/js/unit/Statement.js [hunk 30]
```diff
@@ -74,6 +74,9 @@
                 result = new TinCan.Result(),
                 context = new TinCan.Context(),
                 authority = new TinCan.Agent({ mbox: "tincanjs-test-authority@tincanapi.com" }),
+                attachments = [
+                        new TinCan.Attachment({ usageType: "http://id.tincanapi.com/attachment/test-attachment" })
```
Comment (brianjmiller) [inline test/js/unit/Statement.js]: Whitespace.

File: test/single/Attachment.html [hunk 31]
```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE html>
+<!--
+    Copyright 2014 Rustici Software
```
Comment (brianjmiller) [inline test/single/Attachment.html]: +1

**Comment by FishHooks** [comment]
This last commit adds tests for the new LRS methods, as well as sending and receiving statements with attachments.

Additionally, I've removed all manual header parsing and used the built-in methods of XHR to handle the response headers.

File: src/LRS.js [hunk 32]
```diff
@@ -456,12 +449,34 @@ TinCan client library
                     statementId: stmtId
                 }
             };
+            if (cfg.params.attachments) {
+                requestCfg.params.attachments = true;
+            }
             if (typeof cfg.callback !== "undefined") {
                 callbackWrapper = function (err, xhr) {
-                    var result = xhr;
+                    var result = xhr.responseText,
```
Comment (brianjmiller) [inline src/LRS.js]: Changing the initialization of `result` here breaks the public interface contract in the case of errors. (Which I honestly thought would have been caught by the unit tests :-(). See `cfg.callback(err, result)` below.

File: src/LRS.js [hunk 33]
```diff
@@ -456,12 +449,34 @@ TinCan client library
...
+                        }
+                        else {
+                            boundary = xhr.getResponseHeader("Content-Type").split("boundary=")[1];
+                            result = result.split("--" + boundary + "--")[0].split("--" + boundary);
+
+                            statement = JSON.parse(result[1].split("\r\n\r\n")[1]);
+
+                            for (i = 2; i < result.length; i += 1) {
+                                section = result[i].split("\r\n\r\n");
+                                statement = TinCan.LRS.prototype._assignAttachmentContent({statements: [statement]}, section[0].split("\r\n"), section[1].trim()).statements[0];
```
Comment (brianjmiller) [inline src/LRS.js]: Shouldn't call this method on the prototype without doing so through the instance itself. It should be made either local and not available externally, "static" on the "class", or make the instance available likely via a closure and call it as an instance method.

File: src/LRS.js [hunk 34]
```diff
@@ -711,25 +712,25 @@ TinCan client library
 
             if (typeof cfg.callback !== "undefined") {
                 callbackWrapper = function (err, xhr) {
-                    var result = xhr,
-                        responseLines = xhr.responseText.split("\r\n"),
+                    var result = xhr.responseText,
```
Comment (brianjmiller) [inline src/LRS.js]: +1 See comment above.

File: src/LRS.js [hunk 35]
```diff
@@ -711,25 +712,25 @@ TinCan client library
...
-                                    break;
-                                }
+                            boundary = xhr.getResponseHeader("Content-Type").split("boundary=")[1];
+                            result = result.split("--" + boundary + "--")[0].split("--" + boundary);
+
+                            statements = JSON.parse(result[1].split("\r\n\r\n")[1]);
+
+                            for (i = 2; i < result.length; i += 1) {
+                                section = result[i].split("\r\n\r\n");
+                                statements = TinCan.LRS.prototype._assignAttachmentContent(statements, section[0].split("\r\n"), section[1].trim());
```
Comment (brianjmiller) [inline src/LRS.js]: +1 See comment above.

File: src/LRS.js [hunk 36]
```diff
@@ -887,6 +888,42 @@ TinCan client library
         },
 
         /**
+        Assigns attachment content to the correct attachment to create a StatementsResult object that is sent
+        to the callback of queryStatements()
+
+        @method _assignAttachmentContent
+        @private
+        @param {Array} [stmts] Array of TinCan Statement JSON objects
```
Comment (brianjmiller) [inline src/LRS.js]: "Tin Can" or "TinCan.Statement" but not "TinCan".

File: src/LRS.js [hunk 37]
```diff
@@ -887,6 +888,42 @@ TinCan client library
...
+        @param {Array} [headers] Array containing section headers from the HTTP response
+        @param {String} [content] The content to place into its attachment
+        @return {Array} Array of TinCan Statement JSON objects with correctly assigned attachment content
```
Comment (brianjmiller) [inline src/LRS.js]: +1

File: src/LRS.js [hunk 38]
```diff
@@ -887,6 +888,42 @@ TinCan client library
...
+        */
+        _assignAttachmentContent: function (stmts, headers, content) {
+            var hashToMatch,
+                i,
+                j;
+
+            for (i = 0; i < headers.length; i += 1) {
+                if (headers[i].indexOf("X-Experience-API-Hash") >= 0) {
+                    hashToMatch = headers[i].split(":")[1].trim();
```
Comment (brianjmiller) [inline src/LRS.js]: Still header parsing. What happens if there is whitespace around the ":"?

File: src/LRS.js [hunk 39]
```diff
@@ -887,6 +888,42 @@ TinCan client library
...
+                }
+            }
+
+            for (i = 0; i < stmts.statements.length; i += 1) {
+                if (stmts.statements[i].hasOwnProperty("attachments") && stmts.statements[i].attachments !== null) {
```
Comment (brianjmiller) [inline src/LRS.js]: No need to check if a statement has an attachments property, it always should.

Comment (brianjmiller) [inline src/LRS.js]: Unless these are the JSON objects, have they already been turned into `TinCan.Statement` objects?

File: src/LRS.js [hunk 40]
```diff
@@ -887,6 +888,42 @@ TinCan client library
...
+        to the callback of queryStatements()
+
+        @method _assignAttachmentContent
+        @private
+        @param {Array} [stmts] Array of TinCan Statement JSON objects
+        @param {Array} [headers] Array containing section headers from the HTTP response
+        @param {String} [content] The content to place into its attachment
+        @return {Array} Array of TinCan Statement JSON objects with correctly assigned attachment content
+        */
+        _assignAttachmentContent: function (stmts, headers, content) {
```
Comment (brianjmiller) [inline src/LRS.js]: I wonder if it makes more sense to collect all attachments based on hash, create a map of them, then associate with the statements via the map which should be faster because you'll only run through the statements + attachment lists once, with the current implementation you have to loop over the full list of attachments for each attachment. Granted the lists are small, but no reason to include the overhead AFAICT.

File: src/Statement.js [hunk 41]
```diff
@@ -398,6 +395,29 @@ TinCan client library
...
+        hasAttachmentWithContent: function () {
+            this.log("hasAttachmentWithContent");
+            var i;
+
+            if (! this.hasOwnProperty("attachments")) {
+                return false;
+            }
+            else {
+                for (i = 0; i < this.attachments.length; i += 1) {
+                    if (this.attachments[i].hasOwnProperty("content") && this.attachments[i].content !== null) {
```
Comment (brianjmiller) [inline src/Statement.js]: An `TinCan.Attachment` always has "content" as a property.

File: test/js/unit/LRS.js [hunk 42]
```diff
@@ -189,6 +189,45 @@
...
+                seen[list[i]] = true;
+            }
+            ok(noDupe, "no duplicates in 500");
+        }
+    );
+    test(
+        "_createStatementSegment",
+        function () {
+            var testString = "--testBoundary\r\nContent-Type: application/json\r\n\r\n\"test\"\r\n";
+            deepEqual(TinCan.LRS.prototype._createStatementSegment("testBoundary", "test"), testString, "Statement segment created correctly");
```
Comment (brianjmiller) [inline test/js/unit/LRS.js]: +1 prototype usage.

File: test/js/unit/LRS.js [hunk 43]
```diff
@@ -189,6 +189,45 @@
...
+        }
+    );
+    test(
+        "_createAttachmentSegment",
+        function () {
+            var testString = "--testBoundary\r\nContent-Type: text/plain\r\nContent-Transfer-Encoding: binary\r\nX-Experience-API-Hash: testHash\r\n\r\ntest\r\n";
+            deepEqual(TinCan.LRS.prototype._createAttachmentSegment("testBoundary", "test", "testHash", "text/plain"), testString, "Statement segment created correctly");
```
Comment (brianjmiller) [inline test/js/unit/LRS.js]: +1 prototype usage.

File: test/js/unit/TinCan-async.js [hunk 44]
```diff
@@ -180,7 +242,7 @@
                             id: "http://adlnet.gov/expapi/verbs/attempted"
                         },
                         target: {
-                            id: "http://tincanapi.com/TinCanJS/Test/TinCan_getStatement/async/" + v
+                            id: "http://tincanapi.com/TinCanJS/Test/TinCan_getStatements/async/" + v
```
Comment (brianjmiller) [inline test/js/unit/TinCan-async.js]: This is testing `getStatement`, shouldn't change?

File: test/js/unit/TinCan-async.js [hunk 45]
```diff
@@ -229,11 +291,139 @@
...
+                sendResult = session[v].sendStatement(
+                    {
+                        actor: {
+                            mbox: actorMbox
+                        },
+                        verb: {
+                            id: "http://adlnet.gov/expapi/verbs/attempted"
+                        },
+                        target: {
+                            id: "http://tincanapi.com/TinCanJS/Test/TinCan_getStatements/async/" + v
```
Comment (brianjmiller) [inline test/js/unit/TinCan-async.js]: +1 testing `getStatement`?

File: test/js/unit/TinCan-async.js [hunk 46]
```diff
@@ -229,11 +291,139 @@
...
+                sendResult = session[v].sendStatement(
+                    {
+                        actor: {
+                            mbox: actorMbox
+                        },
+                        verb: {
+                            id: "http://adlnet.gov/expapi/verbs/attempted"
+                        },
+                        target: {
+                            id: "http://tincanapi.com/TinCanJS/Test/TinCan_getStatement/async/" + v
```
Comment (brianjmiller) [inline test/js/unit/TinCan-async.js]: This time testing `getStatements`?

File: test/js/unit/TinCan-async.js [hunk 47]
```diff
@@ -229,11 +291,139 @@
...
+                            function (err, statement) {
+                                start();
+
+                                // clear the "stored" and "authority" properties since we couldn't have known them ahead of time
+                                // TODO: should we check the authority?
+                                statement.stored = null;
+                                statement.authority = null;
+
+                                deepEqual(err, null, "callback: err argument (" + v + ")");
+                                deepEqual(sentStatement.attachments[0], new TinCan.Attachment(statement.attachments[0]), "callback: statement matches (" + v + ")");
```
Comment (brianjmiller) [inline test/js/unit/TinCan-async.js]: Shouldn't  the statement coming back already have a `TinCan.Attachment`, why construct it here?

File: test/js/unit/TinCan-async.js [hunk 48]
```diff
@@ -229,11 +291,139 @@
...
+
+                        var getResult = session[v].getStatements(
+                            {
+                                params: {
+                                    attachments: true
+                                },
+                                callback: function (err, statement) {
+                                    start();
+                                    deepEqual(err, null, "callback: err argument (" + v + ")");
+                                    deepEqual(sentStatement.attachments[0], new TinCan.Attachment(statement.statements[0].attachments[0]), "callback: statement matches (" + v + ")");
```
Comment (brianjmiller) [inline test/js/unit/TinCan-async.js]: +1 to construction question.

File: src/LRS.js (hunk 39 above)
Comment (FishHooks) [inline src/LRS.js]: @brianjmiller At this point they're just JSON, and then I turn them into to `TinCan.Statement` objects after this function returns.

File: src/LRS.js [hunk 49]
```diff
@@ -454,27 +455,35 @@ TinCan client library
             }
             if (typeof cfg.callback !== "undefined") {
                 callbackWrapper = function (err, xhr) {
-                    var result = xhr.responseText,
+                    var result,
```
Comment (brianjmiller) [inline src/LRS.js]: This didn't put the original back so is still broken, `result` needs to be initialized to `xhr` so that it carries through in the callback in the case of errors.

File: src/LRS.js [hunk 50]
```diff
@@ -454,27 +455,35 @@ TinCan client library
...
                             boundary = xhr.getResponseHeader("Content-Type").split("boundary=")[1];
-                            result = result.split("--" + boundary + "--")[0].split("--" + boundary);
+                            result = xhr.responseText.split("--" + boundary + "--")[0].split("--" + boundary);
 
                             statement = JSON.parse(result[1].split("\r\n\r\n")[1]);
 
                             for (i = 2; i < result.length; i += 1) {
                                 section = result[i].split("\r\n\r\n");
-                                statement = TinCan.LRS.prototype._assignAttachmentContent({statements: [statement]}, section[0].split("\r\n"), section[1].trim()).statements[0];
+                                headers = section[0].split("\r\n");
```
Comment (brianjmiller) [inline src/LRS.js]: I'm not convinced any of these `split`s will actually work as intended. They'll handle the simple case correctly, but will fall down on either complex headers, and/or bodies that include non-normal whitespace. Minimally they need to include the second limit argument, specifying that the split should only occur once. For instance, headers can be split across lines or bodies might include multiple line breaks in a row, hence the separation of header from body + boundary.

File: src/LRS.js [hunk 51]
```diff
@@ -454,27 +455,35 @@ TinCan client library
...
+                                for (i = 0; i < headers.length; i += 1) {
+                                    if (headers[i].indexOf("X-Experience-API-Hash") >= 0) {
+                                        attachmentMap[TinCan.Utils.getXAPIHashFromHeader(headers[i])] = section[1].replace(/^\s+|\s+$/g,"");
```
Comment (brianjmiller) [inline src/LRS.js]: This `.replace` is the reason I was concerned about the above as well, this will effectively trim valid whitespace from the body of the content when it shouldn't. We can't rely on the body of the message part not having leading/trailing whitespace, which is the point of the boundary conditions. We likely need to set up some more advanced test cases to make sure we not mangling the attachments.

File: src/LRS.js [hunk 52]
```diff
@@ -893,34 +917,25 @@ TinCan client library
...
+                    for (j = 0; j < stmts[i].attachments.length; j += 1) {
+                        if (content.hasOwnProperty(stmts[i].attachments[j].sha2)) {
+                            stmts[i].attachments[j].content = content[stmts[i].attachments[j].sha2];
                         }
                     }
                 }
             }
 
-            return stmts;
+            return (stmts.length === 1 ? stmts[0] : stmts);
```
Comment (brianjmiller) [inline src/LRS.js]: I think this should be a void return, no reason to return the stmts since they are modified by the function itself.

File: src/LRS.js [hunk 53]
```diff
@@ -893,34 +917,25 @@ TinCan client library
 
         @method _assignAttachmentContent
         @private
-        @param {Array} [stmts] Array of TinCan Statement JSON objects
-        @param {Array} [headers] Array containing section headers from the HTTP response
-        @param {String} [content] The content to place into its attachment
-        @return {Array} Array of TinCan Statement JSON objects with correctly assigned attachment content
+        @param {Array} [stmts] Array of TinCan.Statement JSON objects
+        @param {Hashmap} [content] Hashmap of the content to place into its attachment
```
Comment (brianjmiller) [inline src/LRS.js]: Is "Hashmap" a valid parameter type in these docs? I think we can stick with object in the JS world.

File: src/LRS.js [hunk 54]
```diff
@@ -893,34 +917,25 @@ TinCan client library
...
+        @return {Array} Array of TinCan.Statement JSON objects with correctly assigned attachment content
         */
-        _assignAttachmentContent: function (stmts, headers, content) {
-            var hashToMatch,
-                i,
+        _assignAttachmentContent: function (stmts, content) {
```
Comment (brianjmiller) [inline src/LRS.js]: "content" should probably be renamed, "attachments" or even "attachmentMap" is more clear.

**Comment by brianjmiller** [comment]
Based on the other PR going in, this one needs to be merged/rebased with master again to resolve the conflicts. I'd probably lean rebase.

**Comment by FishHooks** [comment]
Alright. It's refactored, rebased, and retested. I ported the multipart response parsing helper methods straight over from PHP, so it should have everything. The only thing I removed was that PHP had it set up to put the status in the first element of the multipart sections array, so I removed that (I'm guessing it was there because of the way PHP handles the response).

**Comment by tseabrooks** [comment]
@brianjmiller We need to have this merged.

Thanks, TJ

**Comment by brianjmiller** [comment]
Closing in favor of #133.
